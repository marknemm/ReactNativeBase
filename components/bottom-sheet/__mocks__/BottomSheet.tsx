import type { BottomSheetProps } from '@components/bottom-sheet/BottomSheet.interfaces';

const MockBottomSheet = jest.fn().mockImplementation((props: BottomSheetProps) => {
  const ActualComponent = jest.requireActual('@components/bottom-sheet/BottomSheet').default;
  return <ActualComponent {...props} />;
});

export default MockBottomSheet;
