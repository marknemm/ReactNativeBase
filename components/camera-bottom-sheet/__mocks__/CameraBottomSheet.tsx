import type { CameraBottomSheetProps } from '@components/camera-bottom-sheet/CameraBottomSheet.interfaces';

const MockCameraBottomSheet = jest.fn().mockImplementation((props: CameraBottomSheetProps) => {
  const ActualComponent = jest.requireActual('@components/camera-bottom-sheet/CameraBottomSheet').default;
  return <ActualComponent {...props} />;
});

export default MockCameraBottomSheet;
