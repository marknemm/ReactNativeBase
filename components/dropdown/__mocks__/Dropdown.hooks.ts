/**
 * Mocks the useDropdownItems hook.
 */
export const useDropdownItems = jest.fn().mockImplementation((data) => data);
