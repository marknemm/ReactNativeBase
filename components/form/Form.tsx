import { FormProvider } from 'react-hook-form';
import { View } from 'react-native';
import type { FormProps } from './Form.interface';

/**
 * A component that wraps a form and provides form context to its children.
 *
 * @param props The component {@link FormProps}.
 * @returns The {@link Form} component.
 */
const Form: React.FC<FormProps> = ({ children, form, style, ...restViewProps }) => (
  <FormProvider {...form}>
    <View
      style={style}
      testID="rnb-form-view"
      {...restViewProps}
    >
      {children}
    </View>
  </FormProvider>
);

export type * from './Form.interface';
export default Form;
