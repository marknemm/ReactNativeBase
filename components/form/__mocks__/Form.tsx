import type { FormProps } from '@components/form/Form.interface';

const FormMock = jest.fn().mockImplementation((props: FormProps) => {
  const ActualComponent = jest.requireActual('@components/form/Form').default;
  return <ActualComponent {...props} />;
});

export default FormMock;
