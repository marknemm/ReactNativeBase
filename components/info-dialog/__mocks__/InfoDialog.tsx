import type { InfoDialogProps } from '@components/info-dialog/InfoDialog.interfaces';

const MockInfoDialog = jest.fn().mockImplementation((props: InfoDialogProps) => {
  const ActualComponent = jest.requireActual('@components/info-dialog/InfoDialog').default;
  return <ActualComponent {...props} />;
});

export default MockInfoDialog;
