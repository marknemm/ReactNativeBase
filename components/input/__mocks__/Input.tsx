import type { InputProps } from '@components/input/Input.interfaces';

const MockInput = jest.fn().mockImplementation((props: InputProps) => {
  const ActualComponent = jest.requireActual('@components/input/Input').default;
  return <ActualComponent {...props} />;
});

export default MockInput;
