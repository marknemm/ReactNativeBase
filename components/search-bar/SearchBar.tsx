import ActivityIndicator from '@components/activity-indicator/ActivityIndicator';
import Input from '@components/input/Input';
import ClearButton from '@components/input/clear-button/ClearButton';
import { SEARCH_ICON } from '@constants/icons';
import { useFormFieldValue } from '@hooks/form-hooks';
import { useGeneralStyles } from '@hooks/styles-hooks';
import { useColors } from '@hooks/theme-hooks';
import { RefObject, forwardRef, useMemo, useRef } from 'react';
import { View } from 'react-native';
import type { InputRefType, SearchBarFC, SearchBarProps } from './SearchBar.interfaces';
import { useStyles } from './SearchBar.styles';

/**
 * An {@link Input} field that is specialized for search queries.
 *
 * @param props The component {@link SearchBarProps}.
 * @param ref The component reference.
 * @returns The {@link SearchBar} component.
 */
const SearchBar: SearchBarFC = forwardRef(({
  showLoading = false,
  ...inputProps
}, ref) => {
  const value = useFormFieldValue(inputProps);
  const colors = useColors();
  const generalStyles = useGeneralStyles();
  const styles = useStyles(inputProps);
  const inputRef = useRef<InputRefType>();
  ref ??= inputRef;

  const searchIcon = useMemo(() => ({
    ...SEARCH_ICON,
    color: colors.placeholder,
    size: 18,
  }), [colors]);

  const rightIcon = useMemo(() => (
    <View style={generalStyles.view.row}>
      <ActivityIndicator
        isVisible={showLoading}
        style={styles.loading}
        testID="rnb-search-bar-loading"
      />

      <ClearButton
        inputRef={ref as RefObject<InputRefType>}
        isVisible={!!value}
      />
    </View>
  ), [generalStyles, ref, showLoading, styles, value]);

  return (
    <Input
      accessibilityRole="search"
      leftIcon={searchIcon}
      placeholder="Search..."
      rightIcon={rightIcon}
      {...inputProps}
      ref={ref}
    />
  );
});

export type * from './SearchBar.interfaces';
export default SearchBar;
