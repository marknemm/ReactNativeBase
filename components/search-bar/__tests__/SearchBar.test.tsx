import Form from '@components/form/Form';
import SearchBar from '@components/search-bar/SearchBar';
import { genMockForm } from '@hooks/__mocks__/form-hooks';
import AppProvider from '@test/contexts/app/AppProvider';
import { render, screen, userEvent } from '@testing-library/react-native';

describe('<SearchBar />', () => {
  describe('clear button', () => {
    it('does not show clear button when input is empty', () => {
      const form = genMockForm({
        search: '',
      });
      const onChangeText = jest.fn();
      render(
        <Form form={form}>
          <SearchBar
            name="search"
            onChangeText={onChangeText}
          />
        </Form>,
        { wrapper: AppProvider }
      );

      const clearButton = screen.queryByRole('button', { name: 'Clear' });
      expect(clearButton).toBeFalsy();
    });

    it('shows clear button when input is not empty', () => {
      const form = genMockForm({
        search: 'test',
      });
      const onChangeText = jest.fn();
      render(
        <Form form={form}>
          <SearchBar
            name="search"
            onChangeText={onChangeText}
          />
        </Form>,
        { wrapper: AppProvider }
      );

      const clearButton = screen.getByRole('button', { name: 'Clear' });
      expect(clearButton).toBeVisible();
    });

    it('clears the text field on clear button press', async () => {
      const typedStr = 'test';
      const form = genMockForm({
        search: '',
      });
      const onChangeText = jest.fn();
      render(
        <Form form={form}>
          <SearchBar
            name="search"
            onChangeText={onChangeText}
          />
        </Form>,
        { wrapper: AppProvider }
      );

      const input = screen.getByPlaceholderText('Search...');

      await userEvent.type(input, typedStr);
      expect(onChangeText).toHaveBeenLastCalledWith(typedStr);
      expect(form.getValues().search).toBe(typedStr);

      const clearButton = screen.getByRole('button', { name: 'Clear' });

      await userEvent.press(clearButton);
      expect(onChangeText).toHaveBeenLastCalledWith('');
      expect(form.getValues().search).toBe('');
    });
  });

  describe('loading', () => {
    it('does not show loading indicator by default', () => {
      render(
        <SearchBar />,
        { wrapper: AppProvider }
      );

      const loadingIndicator = screen.queryByTestId('rnb-search-bar-loading');
      expect(loadingIndicator).toBeFalsy();
    });

    it('shows loading indicator when showLoading is true', () => {
      render(
        <SearchBar showLoading />,
        { wrapper: AppProvider }
      );

      const loadingIndicator = screen.getByTestId('rnb-search-bar-loading');
      expect(loadingIndicator).toBeVisible();
    });
  });

  describe('snapshots', () => {
    it('renders correctly', () => {
      render(
        <SearchBar />,
        { wrapper: AppProvider }
      );

      expect(screen.toJSON()).toMatchSnapshot();
    });

    it('renders correctly with label, custom placeholder, and clear button', () => {
      render(
        <SearchBar
          label="Search"
          placeholder="Search here"
          value="test"
        />,
        { wrapper: AppProvider }
      );

      expect(screen.toJSON()).toMatchSnapshot();
    });
  });
});
