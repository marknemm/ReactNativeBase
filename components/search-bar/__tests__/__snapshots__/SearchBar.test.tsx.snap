// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SearchBar /> snapshots renders correctly 1`] = `
<View
  style={
    {
      "marginBottom": 6.285714285714286,
      "paddingHorizontal": 0,
      "width": "100%",
    }
  }
  testID="RNE__Input__view-wrapper"
>
  <View
    collapsable={false}
    style={
      {
        "alignItems": "center",
        "borderBottomColor": "#bbb",
        "borderBottomWidth": 1,
        "borderColor": "#86939e",
        "flexDirection": "row",
        "height": 98.9730612244898,
        "transform": [
          {
            "translateX": 0,
          },
        ],
      }
    }
  >
    <View
      style={
        {
          "alignItems": "center",
          "height": 40,
          "justifyContent": "center",
          "marginVertical": 4,
          "paddingRight": 4,
        }
      }
    >
      <View
        style={
          {
            "overflow": "hidden",
          }
        }
        testID="RNE__ICON__CONTAINER"
      >
        <View
          accessibilityRole="button"
          android_ripple={
            {
              "borderless": false,
              "color": "rgba(0, 0, 0, 0.3)",
              "radius": -5,
            }
          }
          disabled={false}
          testID="RNE__ICON__CONTAINER_ACTION"
        >
          <View
            style={
              {
                "alignItems": "center",
                "backgroundColor": "transparent",
                "justifyContent": "center",
              }
            }
            testID="RNE__ICON"
          >
            <Text
              allowFontScaling={false}
              selectable={false}
              style={
                [
                  {
                    "color": "#86939e",
                    "fontSize": 18,
                  },
                  {
                    "backgroundColor": "transparent",
                  },
                  {
                    "fontFamily": "material",
                    "fontStyle": "normal",
                    "fontWeight": "normal",
                  },
                  {},
                ]
              }
              testID="RNE__ICON__Component"
            >
              
            </Text>
          </View>
        </View>
      </View>
    </View>
    <TextInput
      accessibilityRole="search"
      editable={true}
      keyboardAppearance="light"
      onChangeText={[Function]}
      placeholder="Search..."
      placeholderTextColor="#86939e"
      style={
        {
          "color": "#242424",
          "flex": 1,
          "fontSize": 23.57142857142857,
          "height": 98.9730612244898,
          "minHeight": 40,
        }
      }
      testID="RNE__Input__text-input"
      underlineColorAndroid="transparent"
    />
    <View
      style={
        {
          "alignItems": "center",
          "height": 40,
          "justifyContent": "center",
          "marginVertical": 4,
          "paddingRight": 4,
        }
      }
    >
      <View
        style={
          {
            "flexDirection": "row",
          }
        }
      />
    </View>
  </View>
  <Text
    style={
      {
        "color": "#ff3b30",
        "fontSize": 12,
        "margin": 5,
      }
    }
  />
</View>
`;

exports[`<SearchBar /> snapshots renders correctly with label, custom placeholder, and clear button 1`] = `
<View
  style={
    {
      "marginBottom": 6.285714285714286,
      "paddingHorizontal": 0,
      "width": "100%",
    }
  }
  testID="RNE__Input__view-wrapper"
>
  <Text
    style={
      {
        "color": "#86939e",
        "fontSize": 23.57142857142857,
        "fontWeight": "bold",
      }
    }
  >
    Search
  </Text>
  <View
    collapsable={false}
    style={
      {
        "alignItems": "center",
        "borderBottomColor": "#bbb",
        "borderBottomWidth": 1,
        "borderColor": "#86939e",
        "flexDirection": "row",
        "height": 98.9730612244898,
        "transform": [
          {
            "translateX": 0,
          },
        ],
      }
    }
  >
    <View
      style={
        {
          "alignItems": "center",
          "height": 40,
          "justifyContent": "center",
          "marginVertical": 4,
          "paddingRight": 4,
        }
      }
    >
      <View
        style={
          {
            "overflow": "hidden",
          }
        }
        testID="RNE__ICON__CONTAINER"
      >
        <View
          accessibilityRole="button"
          android_ripple={
            {
              "borderless": false,
              "color": "rgba(0, 0, 0, 0.3)",
              "radius": -5,
            }
          }
          disabled={false}
          testID="RNE__ICON__CONTAINER_ACTION"
        >
          <View
            style={
              {
                "alignItems": "center",
                "backgroundColor": "transparent",
                "justifyContent": "center",
              }
            }
            testID="RNE__ICON"
          >
            <Text
              allowFontScaling={false}
              selectable={false}
              style={
                [
                  {
                    "color": "#86939e",
                    "fontSize": 18,
                  },
                  {
                    "backgroundColor": "transparent",
                  },
                  {
                    "fontFamily": "material",
                    "fontStyle": "normal",
                    "fontWeight": "normal",
                  },
                  {},
                ]
              }
              testID="RNE__ICON__Component"
            >
              
            </Text>
          </View>
        </View>
      </View>
    </View>
    <TextInput
      accessibilityRole="search"
      editable={true}
      keyboardAppearance="light"
      onChangeText={[Function]}
      placeholder="Search here"
      placeholderTextColor="#86939e"
      style={
        {
          "color": "#242424",
          "flex": 1,
          "fontSize": 23.57142857142857,
          "height": 98.9730612244898,
          "minHeight": 40,
        }
      }
      testID="RNE__Input__text-input"
      underlineColorAndroid="transparent"
    />
    <View
      style={
        {
          "alignItems": "center",
          "height": 40,
          "justifyContent": "center",
          "marginVertical": 4,
          "paddingRight": 4,
        }
      }
    >
      <View
        style={
          {
            "flexDirection": "row",
          }
        }
      >
        <View
          style={
            [
              {
                "overflow": "hidden",
              },
              {
                "borderRadius": 3.142857142857143,
              },
              undefined,
              false,
            ]
          }
          testID="RNE_BUTTON_WRAPPER"
        >
          <View
            accessibilityLabel="Clear"
            accessibilityRole="button"
            accessibilityState={
              {
                "busy": false,
                "checked": undefined,
                "disabled": false,
                "expanded": undefined,
                "selected": undefined,
              }
            }
            accessibilityValue={
              {
                "max": undefined,
                "min": undefined,
                "now": undefined,
                "text": undefined,
              }
            }
            accessible={true}
            collapsable={false}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              {
                "opacity": 1,
              }
            }
          >
            <View
              style={
                {
                  "alignItems": "center",
                  "backgroundColor": "transparent",
                  "borderColor": "#007aff",
                  "borderRadius": 3.142857142857143,
                  "borderWidth": 0,
                  "flexDirection": "row",
                  "justifyContent": "center",
                  "padding": 6.285714285714286,
                  "paddingHorizontal": 8.285714285714285,
                }
              }
            >
              <View
                style={
                  {
                    "marginHorizontal": 5,
                    "overflow": "hidden",
                  }
                }
                testID="RNE__ICON__CONTAINER"
              >
                <View
                  accessibilityRole="button"
                  android_ripple={
                    {
                      "borderless": false,
                      "color": "rgba(0, 0, 0, 0.3)",
                      "radius": -5,
                    }
                  }
                  disabled={false}
                  testID="RNE__ICON__CONTAINER_ACTION"
                >
                  <View
                    style={
                      {
                        "alignItems": "center",
                        "backgroundColor": "transparent",
                        "justifyContent": "center",
                      }
                    }
                    testID="RNE__ICON"
                  >
                    <Text
                      allowFontScaling={false}
                      selectable={false}
                      style={
                        [
                          {
                            "color": "#86939e",
                            "fontSize": 18,
                          },
                          {
                            "backgroundColor": "transparent",
                          },
                          {
                            "fontFamily": "material",
                            "fontStyle": "normal",
                            "fontWeight": "normal",
                          },
                          {},
                        ]
                      }
                      testID="RNE__ICON__Component"
                    >
                      
                    </Text>
                  </View>
                </View>
              </View>
            </View>
          </View>
        </View>
      </View>
    </View>
  </View>
  <Text
    style={
      {
        "color": "#ff3b30",
        "fontSize": 12,
        "margin": 5,
      }
    }
  />
</View>
`;
